libtenzir/builtins/functions/list.cpp

```diff
@@ -116,6 +118,81 @@ class concatenate : public virtual function_plugin {
   }
 };
 
+class zip final : public function_plugin {
+public:
+  struct arguments {
+    ast::expression left;
+    ast::expression right;
+  };
+
+  auto name() const -> std::string override {
+    return "tql2.zip";
+  }
+
+  auto make_function(invocation inv, session ctx) const
+    -> failure_or<function_ptr> override {
+    auto args = arguments{};
+    TRY(argument_parser2::function("zip")
+          .add(args.left, "<left>")
+          .add(args.right, "<right>")
+          .parse(inv, ctx));
+    return function_use::make([args = std::move(args)](
+                                function_plugin::evaluator eval,
+                                session ctx) -> series {
+      auto left = eval(args.left);
+      auto right = eval(args.right);
+      if (is<null_type>(left.type) and is<null_type>(right.type)) {
+        return series::null(list_type{null_type{}}, left.length());
+      }
+      auto left_list = left.as<list_type>();
+      auto right_list = right.as<list_type>();
+      if (not left_list or not right_list) {
+        if (not left_list) {
+          diagnostic::warning("expected `list`, but got `{}`", left.type.kind())
+            .primary(args.left)
+            .emit(ctx);
+        }
+        if (not right_list) {
+          diagnostic::warning("expected `list`, but got `{}`",
+                              right.type.kind())
+            .primary(args.right)
+            .emit(ctx);
+        }
+        return series::null(list_type{null_type{}}, left.length());
+      }
+      auto builder = series_builder{type{list_type{record_type{
+        {"left", left_list->type.value_type()},
+        {"right", right_list->type.value_type()},
+      }}}};
+      auto left_values = left_list->values();
+      auto right_values = right_list->values();
+      for (auto i = int64_t{0}; i < left.length(); ++i) {
+        auto left_value = check(left_values.next());
+        auto right_value = check(right_values.next());
+        if (not left_value and not right_value) {
+          builder.null();
+          continue;
+        }
+        auto list_builder = builder.list();
+        const auto max_length = std::max(left_value ? left_value->size() : 0,
+                                         right_value ? right_value->size() : 0);
```

Comment:
@jachris brought up that perhaps by default we should warn when the list lengths do not match, introducing an option to choose whether we prefer the min or the max length or warning and returning null.

Reply:
We decided internally to always warn if the lengths of the lists do not match or if either (but not both) list are null. We want to use the max length here in that case.

We can always add options in the future for toggling this behavior, but for now we'll stick with this without options.